### ISSUE TITLE
Non-iterable value num is used in an iterating context

### ISSUE THREAD

**Issue by shenxiangzhuang** [PyCQA/pylint#3105]
### Steps to reproduce
```python
nums = [None] * 10
nums[0] = [1, 2]


def get_subs():
    for num in nums:
        if num is not None:
            for n in num:
                yield n

if __name__ == '__main__':
    print([i for i in get_subs()])
```

### Current behavior

> E1133: Non-iterable value num is used in an iterating context (not-an-iterable)

### Expected behavior
The `if` statement makes sure that the `num` is not `None`, but I still get error message. And I think it is not suitable.

### pylint --version output
>pylint 2.3.1
astroid 2.2.5
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]
(python37)

**Comment by PCManticore** [comment]
Thanks for opening an issue. This is a long standing problem with `pylint`, it cannot figure out that `nums` is an object that allows iteration. We might be able to solve this *at some point* using type annotations, but that's a project too far down the road to be considered planned. I'm going to close this issue as we have already quite a bunch of them that fall under the flow control inference topic.

**Comment by adeharo9** [comment]
I know this issue is closed, but I ran into kind of a similar situation in a much explicit context:

```
for key, value in arguments['required'].items():
        if value is not None:
            if isinstance(value, list):
                for val in value:
                    args.append(val)
            else:
                args.append(value)
```
And even in that context (where I explicitly check for the variable `value` not being `None` and also being of `list` type), it is giving me that error.
